pkg/kubectl/kustomize/k8sdeps/doc.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// It's possible that kustomize's features will be vendored into
+// the kubernetes/kubernetes repo and made available to kubectl
+// commands, while at the same time the kustomize program will
+// continue to exist as an independent CLI.  Vendoring snapshots
+// would be taken just before a kubectl release.
+//
+// This creates a problem in that freestanding-kustomize depends on
+// (for example):
+//
+//   https://github.com/kubernetes/apimachinery/
+//       tree/master/pkg/util/yaml
+//
+// It vendors that package into
+//   sigs.k8s.io/kustomize/vendor/k8s.io/apimachinery/
+//
+// Whereas kubectl-kustomize would have to depend on the "staging"
+// version of this code, located at
+//
+//   https://github.com/kubernetes/kubernetes/
+//       blob/master/staging/src/k8s.io/apimachinery/pkg/util/yaml
+//
+// which is "vendored" via symlinks:
+//   k8s.io/kubernetes/vendor/k8s.io/apimachinery
+// is a symlink to
+//   ../../staging/src/k8s.io/apimachinery
+//
+// The staging version is the canonical, under-development
+// version of the code that kubectl depends on, whereas the packages
+// at kubernetes/apimachinery are periodic snapshots of staging made
+// for outside tools to depend on.
+//
+// apimachinery isn't the only package that poses this problem, just
+// using it as a specific example.
+//
+// The kubectl binary cannot vendor in kustomize code that in
+// turn vendors in the non-staging packages.
+//
+// One way to fix some of this would be to copy code - a hard fork.
+// This has all the problems associated with a hard forking.
+//
+// Another way would be to break the kustomize repo into three:
+//
+// (1) kustomize - repo with the main() function,
+//     vendoring (2) and (3).
+//
+// (2) kustomize-libs - packages used by (1) with no
+//     apimachinery dependence.
+//
+// (3) kustomize-k8sdeps - A thin code layer that depends
+//     on (vendors) apimachinery to provide thin implementations
+//     to interfaces used in (2).
+//
+// The kubectl repo would then vendor from (2) only, and have
```

Comment:
The fact that we have to do this at all is concerning.  Having to write this doc.go out seems like a reason to simplify what this is trying to do.

Reply:
This package is only used for kustomize subcommand. It is not provided for general purpose. Staging may not be a good destination for it.

Reply:
Staging is for all dependencies of kubernetes that are actively developed in tree and are not from a vendor upstream.  Since this looks like a dependency of kubectl that itself cannot depend on kube, vendor or staging is where it belongs.

Reply:
`pkg/kubectl/kustomize/k8sdeps` includes all staging dependencies that are injected to kustomize command. Since it depends on staging libraries, such as api, api-machinery. It can't be put under vendor directory.

We can't put this package under staging either. The staging directories are copied automatically from k/k to public repos, which is the opposite from what we do. The source of truth of `pkg/kubectl/kustomize/k8sdeps` is kustomize repo. It is developed there. For future releases, this package is going to be synced with Kustomize repo before the code-freeze.

Reply:
The canonical source is within the same repo as the vendored kustomize pkg package, right? That is the sort of thing that makes vendoring tools have issues (even if godep happens not to). Do we know what this does to dep, or if it causes issues for go modules?

Reply:
> The canonical source is within the same repo the kustomize pkg package is vendored from, right? 

Yes.
> That is the sort of thing that makes vendoring tools have issues (even if godep happens not to). Do we know what this does to dep, or if it causes issues for go modules?

We didn't try it with dep or go modules. Is there any plan to move from godep to dep or go modules?
When it happens, if this doesn't work. We can split Kustomize into different repos. One is for kustomize/pkg without any staging dependencies and one is for these staging dependencies.

Reply:
I agree with Jordan and Clayton here, we can't have separate vendor tree, even when `kubectl` will be extracted it has to follow the normal rules, so all deps should be part of global `vendor` dir

Reply:
> Is there any plan to move from godep to dep or go modules?

yes. xref https://github.com/kubernetes/kubernetes/issues/68577 https://github.com/kubernetes/kubernetes/issues/63607